src/hotspot/share/gc/shared/referenceProcessor.hpp

```diff
@@ -596,42 +585,15 @@ class ReferenceProcessorAtomicMutator: StackObj {
   }
 };
 
-// This class is an interface used to implement task execution for the
-// reference processing.
-class AbstractRefProcTaskExecutor {
-public:
-
-  // Abstract tasks to execute.
-  class ProcessTask;
-
-  // Executes a task using worker threads.
-  virtual void execute(ProcessTask& task, uint ergo_workers) = 0;
-};
-
-// Abstract reference processing task to execute.
-class AbstractRefProcTaskExecutor::ProcessTask {
-protected:
-  ReferenceProcessor&           _ref_processor;
-  // Indicates whether the phase could generate work that should be balanced across
-  // threads after execution.
-  bool                          _marks_oops_alive;
-  ReferenceProcessorPhaseTimes* _phase_times;
-
-  ProcessTask(ReferenceProcessor& ref_processor,
-              bool marks_oops_alive,
-              ReferenceProcessorPhaseTimes* phase_times)
-    : _ref_processor(ref_processor),
-      _marks_oops_alive(marks_oops_alive),
-      _phase_times(phase_times)
-  { }
+enum class ThreadModel {Multi, Single};
 
+class AbstractClosureContext {
 public:
-  virtual void work(uint worker_id,
-                    BoolObjectClosure& is_alive,
-                    OopClosure& keep_alive,
-                    VoidClosure& complete_gc) = 0;
-
-  bool marks_oops_alive() const { return _marks_oops_alive; }
+  virtual BoolObjectClosure* is_alive(uint worker_id) = 0;
+  virtual OopClosure* keep_alive(uint worker_id) = 0;
+  virtual VoidClosure* complete_gc(uint worker_id) = 0;
+  virtual void prepare_run_task(uint queue_count, ThreadModel tm, bool marks_oops_alive) = 0;
+  uint index(uint id, ThreadModel tm) { return (tm==ThreadModel::Single)?0:id; }
```

Comment:
Please add spaces between operands and operators.

Reply:
I will add spaces.